File: backend/src/app/logs/store.py
"""Persist + prune agent-pushed logfile snapshots.

Only the newest ``KEEP_PER_NAME`` snapshots per ``(instance_id, name)`` are kept;
older ones are dropped on write (and by the scheduled ``prune_logfiles`` safety
net). The pure helpers (``clamp``, ``sanitize_logfiles``, ``surplus_ids``) carry
the logic and are unit-tested; the DB functions are thin wrappers verified live.
"""

from __future__ import annotations

from sqlalchemy import delete, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.db.models import LogEvent, Logfile
from app.logs.events import ExtractedEvent, extract_events

KEEP_PER_NAME = 3
MAX_BYTES = 1_000_000  # per-logfile safety cap (agent tails first, this is a backstop)
_NAME_MAX = 64


def clamp(content: str, max_bytes: int = MAX_BYTES) -> str:
    """Keep only the last ``max_bytes`` characters (newest log lines)."""
    return content[-max_bytes:] if len(content) > max_bytes else content


def sanitize_logfiles(raw: list[dict]) -> list[tuple[str, str]]:
    """Clean an agent ``logfiles`` payload into ``(name, content)`` pairs.

    Drops entries with no name or no content; trims the name and caps content."""
    out: list[tuple[str, str]] = []
    for entry in raw:
        name = str(entry.get("name") or "").strip()[:_NAME_MAX]
        content = entry.get("content") or ""
        if name and content:
            out.append((name, clamp(content)))
    return out


def surplus_ids(ordered_ids: list[int], keep: int = KEEP_PER_NAME) -> list[int]:
    """Given ids newest-first, the ones beyond the keep window (to delete)."""
    return ordered_ids[keep:]


async def _newest_first_ids(session: AsyncSession, instance_id: int, name: str) -> list[int]:
    rows = await session.execute(
        select(Logfile.id)
        .where(Logfile.instance_id == instance_id, Logfile.name == name)
        .order_by(Logfile.collected_at.desc(), Logfile.id.desc())
    )
    return list(rows.scalars().all())


async def record_logfiles(session: AsyncSession, instance_id: int, raw: list[dict]) -> int:
    """Insert pushed snapshots, prune each touched name, refresh critical events."""
    pairs = sanitize_logfiles(raw)
    if not pairs:
        return 0
    for name, content in pairs:
        session.add(
            Logfile(instance_id=instance_id, name=name, bytes=len(content), content=content)
        )
    await session.flush()
    for name in {name for name, _ in pairs}:
        extra = surplus_ids(await _newest_first_ids(session, instance_id, name))
        if extra:
            await session.execute(delete(Logfile).where(Logfile.id.in_(extra)))
    for name, content in pairs:
        await replace_log_events(session, instance_id, name, extract_events(name, content))
    return len(pairs)


async def replace_log_events(
    session: AsyncSession, instance_id: int, log_name: str, events: list[ExtractedEvent]
) -> None:
    """Swap the stored events for one (instance, log) with the latest extraction."""
    await session.execute(
        delete(LogEvent).where(LogEvent.instance_id == instance_id, LogEvent.log_name == log_name)
    )
    for e in events:
        session.add(
            LogEvent(
                instance_id=instance_id,
                log_name=log_name,
                severity=e.severity,
                program=e.program,
                pattern=e.pattern,
                sample=e.sample,
                count=e.count,
                last_ts=e.last_ts,
            )
        )


async def list_logfiles(session: AsyncSession, instance_id: int) -> list[Logfile]:
    rows = await session.execute(
        select(Logfile.id, Logfile.name, Logfile.collected_at, Logfile.bytes)
        .where(Logfile.instance_id == instance_id)
        .order_by(Logfile.name.asc(), Logfile.collected_at.desc())
    )
    return list(rows.all())


async def get_logfile(session: AsyncSession, instance_id: int, logfile_id: int) -> Logfile | None:
    """A single stored snapshot, scoped to the instance (cross-instance ids miss)."""
    row = await session.execute(
        select(Logfile).where(Logfile.id == logfile_id, Logfile.instance_id == instance_id)
    )
    return row.scalars().first()


async def latest_per_name(session: AsyncSession, instance_id: int) -> list[Logfile]:
    """The newest snapshot of each distinct log name for an instance."""
    names = await session.execute(
        select(Logfile.name).where(Logfile.instance_id == instance_id).distinct()
    )
    out: list[Logfile] = []
    for name in sorted(names.scalars().all()):
        row = await session.execute(
            select(Logfile)
            .where(Logfile.instance_id == instance_id, Logfile.name == name)
            .order_by(Logfile.collected_at.desc(), Logfile.id.desc())
            .limit(1)
        )
        latest = row.scalars().first()
        if latest is not None:
            out.append(latest)
    return out


async def backfill_log_events(session: AsyncSession) -> int:
    """Populate an empty log_events table from the newest stored snapshots.

    No-op when events already exist — ingest keeps them current. Returns the
    number of events created."""
    from sqlalchemy import func

    existing = await session.scalar(select(func.count()).select_from(LogEvent))
    if existing:
        return 0
    newest = (
        select(func.max(Logfile.id).label("mid"))
        .group_by(Logfile.instance_id, Logfile.name)
        .subquery()
    )
    rows = await session.execute(select(Logfile).where(Logfile.id.in_(select(newest.c.mid))))
    created = 0
    for lf in rows.scalars():
        events = extract_events(lf.name, lf.content)
        await replace_log_events(session, lf.instance_id, lf.name, events)
        created += len(events)
    return created


async def prune_logfiles(session: AsyncSession) -> int:
    """Keep only KEEP_PER_NAME newest per (instance, name) across all rows."""
    pairs = await session.execute(select(Logfile.instance_id, Logfile.name).distinct())
    deleted = 0
    for instance_id, name in pairs.all():
        extra = surplus_ids(await _newest_first_ids(session, instance_id, name))
        if extra:
            result = await session.execute(delete(Logfile).where(Logfile.id.in_(extra)))
            deleted += result.rowcount or 0
    return deleted

File: orbit/lib/orbit/logs/store.ex
defmodule Orbit.Logs.Store do
  @moduledoc """
  Persist + prune agent-pushed logfile snapshots and refresh the extracted
  critical events. Mirror of backend/src/app/logs/store.py.

  Only the newest `keep_per_name/0` snapshots per (instance_id, name) survive a
  write. The pure helpers (`clamp/1`, `sanitize/1`, `surplus/1`) carry the logic
  and are unit-tested; the DB path is verified live (the test DB is schemaless
  under the Alembic monopoly).
  """

  import Ecto.Query

  alias Orbit.Logs.{Events, LogEvent, Logfile}
  alias Orbit.Repo

  @keep_per_name 3
  @max_bytes 1_000_000
  @name_max 64

  @doc "Newest-snapshots-per-name retention window."
  def keep_per_name, do: @keep_per_name

  @doc "Keep only the last `max` characters (newest log lines)."
  @spec clamp(String.t(), pos_integer()) :: String.t()
  def clamp(content, max \\ @max_bytes) do
    if String.length(content) > max, do: String.slice(content, -max, max), else: content
  end

  @doc "Clean an agent `logfiles` payload into `{name, content}` pairs (drop empty)."
  @spec sanitize([map()]) :: [{String.t(), String.t()}]
  def sanitize(raw) when is_list(raw) do
    for entry <- raw,
        name =
          entry
          |> Map.get("name", "")
          |> to_string()
          |> String.trim()
          |> String.slice(0, @name_max),
        content = to_string(Map.get(entry, "content", "")),
        name != "" and content != "" do
      {name, clamp(content)}
    end
  end

  def sanitize(_), do: []

  @doc "Given ids newest-first, the ones beyond the keep window (to delete)."
  @spec surplus([integer()], pos_integer()) :: [integer()]
  def surplus(ordered_ids, keep \\ @keep_per_name), do: Enum.drop(ordered_ids, keep)

  @doc """
  Insert pushed snapshots, prune each touched name to the keep window, and
  refresh the extracted events. Returns the number of stored snapshots.
  """
  @spec ingest(integer(), [map()]) :: non_neg_integer()
  def ingest(instance_id, raw) do
    case sanitize(raw) do
      [] ->
        0

      pairs ->
        now = DateTime.utc_now() |> DateTime.truncate(:second)

        Enum.each(pairs, fn {name, content} ->
          Repo.insert!(%Logfile{
            instance_id: instance_id,
            name: name,
            bytes: String.length(content),
            content: content,
            collected_at: now
          })
        end)

        pairs |> Enum.map(&elem(&1, 0)) |> Enum.uniq() |> Enum.each(&prune(instance_id, &1))

        Enum.each(pairs, fn {name, content} ->
          replace_events(instance_id, name, Events.extract_events(name, content), now)
        end)

        length(pairs)
    end
  end

  defp prune(instance_id, name) do
    ids =
      Repo.all(
        from(l in Logfile,
          where: l.instance_id == ^instance_id and l.name == ^name,
          order_by: [desc: l.collected_at, desc: l.id],
          select: l.id
        )
      )

    case surplus(ids) do
      [] -> :ok
      extra -> Repo.delete_all(from(l in Logfile, where: l.id in ^extra))
    end
  end

  defp replace_events(instance_id, log_name, events, now) do
    Repo.delete_all(
      from(e in LogEvent, where: e.instance_id == ^instance_id and e.log_name == ^log_name)
    )

    Enum.each(events, fn e ->
      Repo.insert!(%LogEvent{
        instance_id: instance_id,
        log_name: log_name,
        severity: e.severity,
        program: e.program,
        pattern: e.pattern,
        sample: e.sample,
        count: e.count,
        last_ts: e.last_ts,
        updated_at: now
      })
    end)
  end

  @doc "Newest snapshot per name for an instance (metadata; content dropped)."
  @spec latest_per_name(integer()) :: [map()]
  def latest_per_name(instance_id) do
    Repo.all(
      from(l in Logfile,
        where: l.instance_id == ^instance_id,
        order_by: [asc: l.name, desc: l.collected_at, desc: l.id],
        select: %{id: l.id, name: l.name, collected_at: l.collected_at, bytes: l.bytes}
      )
    )
    |> Enum.uniq_by(& &1.name)
  end

  @doc "One snapshot (with content) for an instance, or nil. Scope is the caller's."
  @spec get_logfile(integer(), integer()) :: Logfile.t() | nil
  def get_logfile(instance_id, logfile_id) do
    Repo.one(from(l in Logfile, where: l.instance_id == ^instance_id and l.id == ^logfile_id))
  end

  @doc "All extracted events for an instance, worst-first (severity, then count)."
  @spec list_events(integer()) :: [LogEvent.t()]
  def list_events(instance_id) do
    Repo.all(
      from(e in LogEvent,
        where: e.instance_id == ^instance_id,
        order_by: [asc: e.severity, desc: e.count, asc: e.program]
      )
    )
  end
end
